mmeval/metrics/dota_map.py

```diff
@@ -0,0 +1,294 @@
+                        if area_mask:
+                            fp[iou_thr_idx, area_idx, pred_bbox_idx] = 1
+
+        return tp, fp
+
+    def calculate_class_tpfp(self, predictions: List[dict],
+                             groundtruths: List[dict], class_index: int,
+                             pool: Optional[Pool]) -> Tuple:
+        """Calculate the tp and fp of the given class index.
+
+        Args:
+            predictions (List[dict]): A list of dict. Each dict is the
+                detection result of an image.
+            groundtruths (List[dict]): A list of dict. Each dict is the
+                ground truth of an image.
+            class_index (int): The class index.
+            pool (Optional[Pool]): A instance of :class:`multiprocessing.Pool`.
+                If None, do not use multiprocessing.
+
+        Returns:
+            tuple (tp, fp, num_gts):
+            - tp (numpy.ndarray): Shape (num_ious, num_scales, num_pred),
+              the true positive flag of each predicted bbox for this class.
+            - fp (numpy.ndarray): Shape (num_ious, num_scales, num_pred),
+              the false positive flag of each predicted bbox for this class.
+            - num_gts (numpy.ndarray): Shape (num_ious, num_scales), the
+              number of ground truths.
+        """
+        class_preds = self.get_class_predictions(predictions, class_index)
+        class_gts, class_ignore_gts = self.get_class_gts(
+            groundtruths, class_index)
+        if pool is not None:
+            num_images = len(class_preds)
+            tpfp_list = pool.starmap(
+                self._calculate_image_tpfp,
+                zip(
+                    class_preds,
+                    class_gts,
+                    class_ignore_gts,
+                    [self.iou_thrs] * num_images,
+                    [self._area_ranges] * num_images,
+                ))
+        else:
+            tpfp_list = []
+            for img_idx in range(len(class_preds)):
+                tpfp = self._calculate_image_tpfp(class_preds[img_idx],
+                                                  class_gts[img_idx],
+                                                  class_ignore_gts[img_idx],
+                                                  self.iou_thrs,
+                                                  self._area_ranges)
+                tpfp_list.append(tpfp)
+
+        image_tp_list, image_fp_list = tuple(zip(*tpfp_list))
+        sorted_indices = np.argsort(-np.vstack(class_preds)[:, -1])
+        tp = np.concatenate(image_tp_list, axis=2)[..., sorted_indices]
+        fp = np.concatenate(image_fp_list, axis=2)[..., sorted_indices]
+        num_gts = np.zeros((self.num_iou, self.num_scale), dtype=int)
+        for idx, (min_area, max_area) in enumerate(self._area_ranges):
+            area_mask = filter_by_bboxes_area_rotated(
+                np.vstack(class_gts), min_area, max_area)
+            num_gts[:, idx] = np.sum(area_mask)
```

Comment:
The `filter_by_bboxes_area_rotated` seems should be a class method to avoid the overwrite of `calculate_class_tpfp`?